File: hw4/customers.py
import pandas as pd
import numpy as np
import warnings
from loans import tidy_loans
warnings.filterwarnings('ignore')

def tidy_customer(account, districts, links, cards, transactions, payment_orders, loans):
    loans_tidy = tidy_loans(loans)
    combined1 = pd.merge(account, districts, left_on='district_id', right_on='id', how='left').drop(['id_y'], axis=1)
    combined1.rename(columns={'id_x': 'account_id', 'name': 'district_name', 'date': 'open_date'}, inplace=True)
    combined3 = pd.merge(cards, links, left_on='link_id', right_on='id', how='right')
    combined3.rename(columns={'id_x': 'card_id', 'type_x': 'card_type', 'issue_date': 'card_issue_date', 'type_y': 'card_type'}, inplace=True)
    combined3.drop(['id_y'], axis=1, inplace=True)
    card_account = combined3.groupby(['account_id']).count().reset_index()
    links_account = links.groupby(['account_id']).count().reset_index()
    links_account.rename(columns={'client_id': 'num_customers'}, inplace=True)
    combined2 = pd.merge(combined1, links_account, left_on='account_id', right_on='account_id', how='left')
    combined4 = pd.merge(combined2, card_account, left_on='account_id', right_on='account_id', how='left')
    combined4.rename(columns={'card_id': 'credit_cards'}, inplace=True)
    combined5 = pd.merge(combined4, loans_tidy, left_on='account_id', right_on='account_id', how='left')
    combined5.rename(columns={'id_y': 'loan'}, inplace=True)
    combined5['loan'] = combined5['loan'].apply(lambda x: False if np.isnan(x) else True)
    combined5.rename(columns={'amount': 'loan_amount', 'payments': 'loan_payments', 'term': 'loan_term', 'active_status': 'loan_status'}, inplace=True)
    combined5['loan_default'] = combined5['payment'].apply(lambda x: True if x == 'default' else (False if x in ['full', 'in_progress'] else x))
    max_withdrawal = transactions.groupby(['account_id'])['amount'].agg(max)
    min_withdrawal = transactions.groupby(['account_id'])['amount'].agg(min)  
    combined6 = pd.merge(combined5, max_withdrawal, left_on='account_id', right_on='account_id', how='left')
    combined7 = pd.merge(combined6, min_withdrawal, left_on='account_id', right_on='account_id', how='left')
    combined7.rename(columns={'amount_x': 'max_withdrawal', 'amount_y': 'min_withdrawal'}, inplace=True)
    payment_counts = payment_orders.groupby(['account_id'])['id'].count().reset_index()
    payment_counts.columns = ['account_id', 'cc_payments']
    combined8 = pd.merge(combined7, payment_counts, left_on='account_id', right_on='account_id',how='left')
    max_balance = transactions.groupby(['account_id'])['balance'].agg(max).to_frame().reset_index().rename(columns={'balance': 'max_balance'})
    min_balance = transactions.groupby(['account_id'])['balance'].agg(min).to_frame().reset_index().rename(columns={'balance': 'min_balance'})
    combined9 = pd.concat([combined8, max_balance, min_balance], axis=1)
    final_df = combined9[['account_id', 'district_name', 'open_date', 'statement_frequency', 'num_customers', 'credit_cards', 'loan', 'loan_amount', 'loan_payments', 'loan_term', 'loan_status', 'loan_default', 'max_withdrawal', 'min_withdrawal', 'cc_payments', 'max_balance', 'min_balance']]
    final_df = final_df.iloc[:,2:]
    final_df.to_csv('customers_py.csv')  
    return final_df 
      
if __name__ == '__main__':
    account = pd.read_csv('data/accounts.csv')
    districts = pd.read_csv('data/districts.csv')
    links = pd.read_csv('data/links.csv')
    cards = pd.read_csv('data/cards.csv')
    transactions = pd.read_csv('data/transactions.csv')
    payment_orders = pd.read_csv('data/payment_orders.csv')
    loans = pd.read_csv('data/loans.csv')
    tidy_customer(account, districts, links, cards, transactions, payment_orders, loans)

File: hw4/customers.R
library(magrittr) 
library(dplyr) 
library(tidyverse)
library(reshape)
library(data.table)
source(here::here('loans.R'))

tidy_customers <- function(accounts, districts, links, cards, loans, transactions, payment_orders){
  loans_new <- tidy_loans(loans)
  combined1 <- merge(accounts, districts, all.x=TRUE, by.x = "district_id", by.y = "id")
  setnames(combined1, "id", "account_id")
  setnames(combined1, "name", "district_name")
  setnames(combined1, "date", "open_date")
  combined3 <- merge(cards, links, all.y=TRUE, by.x = "link_id", by.y = "id")
  setnames(combined3, "id", "card_id")
  setnames(combined3, "type.x", "card_type")
  setnames(combined3, "issue_date", "card_issue_date")
  setnames(combined3, "type.y", "customer_type")
  card_account <- combined3 %>%
    group_by(account_id) %>%
    summarise(count = n())
  links_account <- links %>%
    group_by(account_id) %>%
    summarise(num_customers = n())
  combined2 <- merge(combined1, links_account, all.x=TRUE, by.x = "account_id", by.y = "account_id")
  combined4 <- merge(combined2, card_account, all.x=TRUE, by.x='account_id', by.y='account_id')
  setnames(combined4, "count", "credit_cards")
  combined5 <- merge(combined4, loans_new, by.x='account_id', by.y='account_id', all.x =TRUE)
  setnames(combined5, "id", "loan")
  combined5$loan <- ifelse(is.na(combined5$loan), FALSE, TRUE)
  setnames(combined5, "amount", "loan_amount")
  setnames(combined5, "payments", "loan_payments")
  setnames(combined5, "term", "loan_term")
  setnames(combined5, "active_status", "loan_status")
  combined5$loan_default <- ifelse((combined5$payment) == "default", TRUE, ifelse(combined5$payment %in% c("full", "in_progress"), FALSE, NA))
  max_withdrawal <- transactions %>% group_by(account_id) %>% summarize(max = max(amount))
  min_withdrawal <- transactions %>% group_by(account_id) %>% summarize(min = min(amount))
  combined6 <- merge(combined5, max_withdrawal, by.x='account_id', by.y='account_id', all.x=TRUE)
  combined7 <- merge(combined6, min_withdrawal, by.x='account_id', by.y='account_id', all.x=TRUE)
  setnames(combined7, "max", "max_withdrawal")
  setnames(combined7, "min", "min_withdrawal")
  payment_counts = payment_orders %>% group_by(account_id) %>% summarize(cc_payments = n())
  combined8 <- merge(combined7, payment_counts, by.x='account_id', by.y='account_id',all.X=TRUE)
  max_balance <- transactions %>% group_by(account_id) %>% summarize(max_balance = max(balance))
  min_balance <- transactions %>% group_by(account_id) %>% summarize(min_balance = min(balance))
  combined9 <- merge(combined8, max_balance, all.x=TRUE, by.x='account_id', by.y='account_id')
  combined9 <- merge(combined9, min_balance, all.x=TRUE, by.x='account_id', by.y='account_id')
  final_df <- combined9 %>% select(c('account_id', 'district_name', 'open_date', 'statement_frequency', 'num_customers', 'credit_cards', 'loan', 'loan_amount', 'loan_payments', 'loan_term', 'loan_status', 'loan_default', 'max_withdrawal', 'min_withdrawal', 'cc_payments', 'max_balance', 'min_balance'))
  final_df$loan_term <- as.numeric(final_df$loan_term)
  write.csv(final_df, "customers_r.csv")
  return(final_df)
}

main <- function(){
  accounts <- read.csv("data/accounts.csv")
  districts <- read.csv("data/districts.csv")
  links <- read.csv("data/links.csv")
  cards <- read.csv("data/cards.csv")
  loans <- read.csv("data/loans.csv")
  transactions <- read.csv("data/transactions.csv")
  payment_orders <- read.csv("data/payment_orders.csv")
  tidy_customers(accounts, districts, links, cards, loans, transactions, payment_orders)
}

if (!interactive()) {
  main()
}